File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/InventoryQuery/Filters/NameContainsFilterDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.InventoryQuery.Filters;

/// <summary>
///     A filter that filters inventory items by checking if their name contains a substring.
/// </summary>
public sealed class NameContainsFilterDecorator : InventoryQueryDecoratorBase
{
    private readonly string _substring;

    /// <summary>
    ///     Initializes a new instance of the <see cref="NameContainsFilterDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="substring">The substring to search for in the item names.</param>
    /// <exception cref="ArgumentNullException">Thrown when substring is null.</exception>
    public NameContainsFilterDecorator(IInventoryQuery inner, string? substring) : base(inner)
    {
        _substring = substring ?? throw new ArgumentNullException(nameof(substring));
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
        => string.IsNullOrWhiteSpace(_substring) ? null : $"Filter: Name contains \"{_substring}\"";

    /// <inheritdoc />
    protected override IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items)
    {
        if (string.IsNullOrWhiteSpace(_substring))
            return items;

        return items.Where(i =>
                i.Name.Contains(_substring, StringComparison.OrdinalIgnoreCase))
            .ToList()
            .AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/filters/NameContainsFilterDecorator.java
package assignment2solution.domain.inventoryquery.filters;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * A filter that filters inventory items by checking if their name contains a substring.
 */
public final class NameContainsFilterDecorator extends InventoryQueryDecoratorBase {
    private final String substring;

    /**
     * Initializes a new instance of the {@link NameContainsFilterDecorator} class.
     */
    public NameContainsFilterDecorator(IInventoryQuery inner, String substring) {
        super(inner);
        this.substring = Objects.requireNonNull(substring, "substring");
    }

    @Override
    protected String getAppliedDescription() {
        if (substring.trim().isEmpty()) {
            return null;
        }
        return "Filter: Name contains \"" + substring + "\"";
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        if (substring.trim().isEmpty()) {
            return items;
        }

        var lowered = substring.toLowerCase(Locale.ROOT);
        return items.stream()
            .filter(i -> i.getName().toLowerCase(Locale.ROOT).contains(lowered))
            .collect(Collectors.toUnmodifiableList());
    }
}
